### PR TITLE
httpx: fix handling of async hooks

### LINKED ISSUE #2734: Sync hook used as async hook in opentelemetry-instrumentation-httpx causing TypeError
### Describe your environment

OS: Ubuntu
Python 3.10.12
Name: opentelemetry-instrumentation-httpx
Version: 0.46b0

### What happened?

code from lib
```python
request_hook = kwargs.get("request_hook")
response_hook = kwargs.get("response_hook")
async_request_hook = kwargs.get("async_request_hook", request_hook)
async_response_hook = kwargs.get("async_response_hook", response_hook
```
If you set up `request_hook`, it will also be used for `async_request_hook` if you don't initialize `async_request_hook`. This could lead to issues because you don't want to pass an async function to `request_hook`. If a non-async function is passed, it may not work as expected. It seems that there needs to be a wrapper around `request_hook` to check if it is not async and then wrap it accordingly. 

### Steps to Reproduce
```python
def hook(span, req):
    pass

HTTPXClientInstrumentor().instrument(request_hook=hook)
    
with Client() as client:
    client.post("http://localhost:9999/hello")
async with AsyncClient() as client:
    await client.post("http://localhost:9999/hello")
```

### Expected Result

Both requests completed

### Actual Result

```
await self._request_hook(span, request_info)
TypeError: object NoneType can't be used in 'await' expression
```

### Additional context

_No response_

### Would you like to implement a fix?

None

### PR DESCRIPTION
**Pull request by xrmx**
# Description

Don't default to sync hooks for async hooks, then check that they are actually async. Needed to fix current tests that were testing only the sync path because tests are inherited from a base class :broken_heart: .

Fixes #2734
Supersedes #2794

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A

# Does This PR Require a Core Repo Change?

- [ ] Yes. - Link to PR: 
- [x] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [ ] Followed the style guidelines of this project
- [x] Changelogs have been updated
- [x] Unit tests have been added
- [ ] Documentation has been updated

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -20,6 +20,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## Fixed
 
+- `opentelemetry-instrumentation-httpx` fix handling of async hooks
+  ([2823](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2823))
```
Comment (emdneto) [inline CHANGELOG.md:24]:
```suggestion
  ([#2823](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2823))
```

File: instrumentation/opentelemetry-instrumentation-httpx/tests/test_httpx_integration.py [hunk 2]
```diff
@@ -1214,3 +1226,36 @@ def test_basic_multiple(self):
...
+            tracer_provider=self.tracer_provider,
+            async_request_hook=_request_hook,
+        )
+        client = self.create_client()
+        result = self.perform_request(self.URL, client=client)
+
+        self.assertEqual(result.text, "Hello!")
+        span = self.assert_span()
+        self.assertEqual(span.name, "GET")
+        HTTPXClientInstrumentor().uninstrument()
```
Comment (emdneto) [inline instrumentation/opentelemetry-instrumentation-httpx/tests/test_httpx_integration.py:1261]: Are we missing out on testing the issue behavior? In cases we have only request_hook, it shouldn't generate an error anymore:

For example, this should fail without your fix:
```
            HTTPXClientInstrumentor().instrument(
                tracer_provider=self.tracer_provider,
                request_hook=_request_hook,
            )
            client = self.create_client()
            result = self.perform_request(self.URL, client=client)

            self.assertEqual(result.text, "Hello!")
            span = self.assert_span()
            self.assertEqual(span.name, "GET" + self.URL)
            HTTPXClientInstrumentor().uninstrument()
```
Will produce the error: 

```
        with self._tracer.start_as_current_span(
            span_name, kind=SpanKind.CLIENT, attributes=span_attributes
        ) as span:
            exception = None
            if callable(self._request_hook):
>               await self._request_hook(span, request_info)
E               TypeError: object NoneType can't be used in 'await' expression
```

Comment (xrmx) [reply instrumentation/opentelemetry-instrumentation-httpx/tests/test_httpx_integration.py:1261]: We can be more defensive but I think the regression was to use a sync function in the async path under the hood. Given that we have two different parameters for sync and async hooks using a sync function with an async client I think is a progamming error more than an issue of the instrumentation.

Comment (xrmx) [reply instrumentation/opentelemetry-instrumentation-httpx/tests/test_httpx_integration.py:1261]: Maybe we can add a check when using it too

Comment (emdneto) [reply instrumentation/opentelemetry-instrumentation-httpx/tests/test_httpx_integration.py:1261]: Right. Taking a second look at the tests I'm fine with

**Review by emdneto — APPROVED** [review]

**Pull request merged**